rust/user-info-fetcher/src/backend/keycloak.rs

```diff
@@ -0,0 +1,180 @@
+///
+/// [users]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#_get_adminrealmsrealmusers
+/// [user-by-id]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#_get_adminrealmsrealmusersid
+/// [UserRepresentation]: https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#UserRepresentation
+/// [issue-20292]: https://github.com/keycloak/keycloak/issues/20294
+#[derive(Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct UserMetadata {
+    id: String,
+    username: String,
+    #[serde(default)]
+    attributes: HashMap<String, Vec<String>>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GroupMembership {
+    path: String,
+}
+
+pub(crate) async fn get_user_info(
+    req: &UserInfoRequest,
+    http: &reqwest::Client,
+    credentials: &Credentials,
+    config: &crd::KeycloakBackend,
+) -> Result<UserInfo, Error> {
+    let crd::KeycloakBackend {
+        client_credentials_secret: _,
+        admin_realm,
+        user_realm,
+        hostname,
+        port,
+        root_path,
+        tls,
+    } = config;
+
+    // We re-use existent functionality from operator-rs, besides it being a bit of miss-use.
+    // Some attributes (such as principal_claim) are irrelevant, and will not be read by the code-flow we trigger.
+    let wrapping_auth_provider = oidc::AuthenticationProvider::new(
+        hostname.clone(),
+        *port,
+        root_path.clone(),
+        tls.clone(),
+        String::new(),
+        Vec::new(),
+        None,
+    );
+    let keycloak_url = wrapping_auth_provider
+        .endpoint_url()
+        .context(ParseOidcEndpointUrlSnafu)?;
+
+    let authn = send_json_request::<OAuthResponse>(
+        http.post(
+            keycloak_url
+                .join(&format!(
+                    "realms/{admin_realm}/protocol/openid-connect/token"
+                ))
+                .context(ConstructOidcEndpointPathSnafu)?,
+        )
+        .basic_auth(&credentials.username, Some(&credentials.password))
+        .form(&[("grant_type", "client_credentials")]),
+    )
+    .await
+    .context(LogInSnafu)?;
+
+    let users_base_url = keycloak_url
+        .join(&format!("admin/realms/{user_realm}/users/"))
+        .context(ConstructOidcEndpointPathSnafu)?;
+
+    let user_info = match req {
+        UserInfoRequest::UserInfoRequestById(req) => {
+            let user_id = req.id.clone();
+            send_json_request::<UserMetadata>(
+                http.get(
+                    users_base_url
+                        .join(&req.id)
+                        .context(ConstructOidcEndpointPathSnafu)?,
+                )
+                .bearer_auth(&authn.access_token),
+            )
+            .await
+            .context(UserNotFoundByIdSnafu { user_id })?
+        }
+        UserInfoRequest::UserInfoRequestByName(req) => {
```

Comment:
Did we have any application that actually must query by username?

Reply:
If we do keep this around, should be `UserInfoRequestByUsername`.

Reply:
> Did we have any application that actually must query by username?

Yes, Trino will when Superset is being used, because Superset is only sending through the `preferred_username` (when using OIDC).

Reply:
Are we sure that Superset is sending the `preferred_username` (IdP username) and not the Superset username?

Reply:
You're right, my description wasn't accurate. 
Superset it sending through the username as it appears in the Superset user DB. When using OIDC with Keycloak, this comes from the `preferred_username` claim.

Trino is then sending whatever that is (it comes from the `x-trino-username` header).

Reply:
Right, so are we sure that these always match up? What happens when migrating a Superset from internal authn to OIDC? What happens if the username specified by `preferred_username` is already taken?

Reply:
ignore this message... github bug that keeps moving focus to this :disappointed:

Reply:
> Are we sure that these always match up?

When Impersonation is enabled on the Database Connection in Superset, Superset will send the logged in username in the `z-trino-username` header. When using OIDC, this originates from a claim, and in the case of Keycloak, is [hardcoded as `preferred_username`](https://github.com/dpgaspar/Flask-AppBuilder/blob/ba63c5c1629101dadf43e68a373402ee116f6ba9/flask_appbuilder/security/manager.py#L659-L671).

> What happens when migrating a Superset from internal authn to OIDC?
> What happens if the username specified by preferred_username is already taken?

No idea, my assumption would be as long as the applicable JWT claim (in the case of Keycloak, the `preferred_username` claim) matches the existing user, then it _is_ the same user. Just different authn mechanism. 

Would be interesting to test, but I believe that to be out-of-scope for this PR.

Reply:
> When using OIDC, this originates from a claim, and in the case of Keycloak, is [hardcoded as preferred_username](https://github.com/dpgaspar/Flask-AppBuilder/blob/ba63c5c1629101dadf43e68a373402ee116f6ba9/flask_appbuilder/security/manager.py#L659-L671).

I'll need to look deeper, but often that sort of thing is just the _suggested_ data to prefill the sign-up form. And as you can see, in that file it's just as often [hard-coded to *not* match the backend username](https://github.com/dpgaspar/Flask-AppBuilder/blob/ba63c5c1629101dadf43e68a373402ee116f6ba9/flask_appbuilder/security/manager.py#L645).

> Would be interesting to test, but I believe that to be out-of-scope for this PR.

Yes and no. I'd rather not add an API path that encourages people to do the wrong thing... :/

Reply:
Yeah, I have no idea what is going on with that openshift user example. But that seems far beyond out control.

> I'd rather not add an API path that encourages people to do the wrong thing... :/

Sure, but I think we are at the mercy of the product implementations. Superset uses Flask-Appbuilder with the already mentioned oddities baked in. Trino only allows a single piece of data to be sent for impersonation. I'm not really sure there is much we can do on our side to discourage what we think is the wrong thing.

Do you have any suggestions?

Reply:
Update: we had a discussion to give out opinions and try to reach a resolution.

The outcome of it was:
- Merge as is, and (optionally?) create a new issue for any renaming of fields (for example, `username` -> `effective_username` in the UIF response data).
- Leave docs hidden for now. Merging this now is not a promise of stability or release (even if the code is technically part of the next release because it is in the main branch).
- Have a discussion about branching philosophy as there appears to be multiple trains of thought on it, each with their own advantages and disadvantages.

_I hope I have captured this correctly, but I will leave it open for @nightkr to confirm, or add/correct anything I got wrong._

Reply:
That's my understanding as well.

Reply:
Cool, I'll mark as resolved and merge it.